### PR TITLE
SYNPY-537 copyFileHandles

### PR DESCRIPTION
**Pull request by minneker**

### CODE REVIEW

File: synapseclient/.synapseConfig [hunk 1]
```diff
@@ -1,14 +1,19 @@
...
 
 ## Used for logging in to Synapse
-## Alternatively, you can use rememberMe=True in synapseclient.login or login subcommand of the commandline client.
+## you may also specify an apikey instead of password. If both password and apikey are specified, the apikey is ignored
+## Alternatively you can use rememberMe=True in synapseclient.login or login subcommand of the commandline client to 
+## cache your API key elsewhere.
 #[authentication]
-#username = <username>
+username ="minneker"
+password ="ilSkate!21"
```
Comment (thomasyu888) [inline synapseclient/.synapseConfig]: Change your password immediately.

File: synapseclient/client.py [hunk 2]
```diff
@@ -3064,6 +3065,75 @@ def sendMessage(self, userIds, messageSubject, messageBody, contentType="text/pl
             fileHandleId=fileHandleId)
         return self.restPOST(uri='/message', body=json.dumps(message))
 
+    ############################################################
+    #                       File Handles                       #
```
Comment (kimyen) [inline synapseclient/client.py]: This should be removed.

File: synapseutils/copy.py [hunk 3]
```diff
@@ -4,48 +4,75 @@
...
 import re
 import json
+import itertools
 ############################################################
 #                  Copy Functions                          #
 ############################################################
 
 
-def copyFileHandles(syn, fileHandles, associateObjectTypes, associateObjectIds, contentTypes, fileNames):
+def copyFileHandles(self, file_handles, obj_types, obj_ids, con_types=None, file_names=None):
```
Comment (kimyen) [inline synapseutils/copy.py]: keep the api parameter name

File: tests/unit/synapseutils/unit_test_synapseutils_copy.py [hunk 4]
```diff
@@ -41,4 +41,56 @@ def test_copyWiki_input_validation():
...
+                                    "fileHandleId": "41414463",
+                                    "associateObjectId": "20459234",
+                                    "associateObjectType": "FileEntity"
+                                },
+                                "newContentType": "wiki",
+                                "newFileName": "test"
+                            }
+                        ]
+                    }
+        self.mock_restPOST.assert_called_once_with('/filehandles/copy', body=json.dumps(expected), endpoint=syn.fileHandleEndpoint)
```
Comment (kimyen) [inline tests/unit/synapseutils/unit_test_synapseutils_copy.py]: test cases:
1. mock that the POST response contains mix results including success and failure
2. mock case where all items fit in 1 call
3. mock case where all items got split into 2 calls, and the results are combined.

File: tests/integration/synapseutils/test_synapseutils_copy.py [hunk 5]
```diff
@@ -309,85 +309,3 @@ def test_entitySubPageId_and_destinationSubPageId(self):
         assert_equals(temp.title, self.sub_subwiki.title)
         assert_equals(temp.markdown, self.sub_subwiki.markdown)
 
-
```
Comment (kimyen) [inline tests/integration/synapseutils/test_synapseutils_copy.py]: add 1 integration test that make copy of 2 file handles

Comment (kimyen) [reply tests/integration/synapseutils/test_synapseutils_copy.py]: recommandation:
1. in the test setup, create an external file handle (URL to files on the internet) (you can use this URL: https://www.synapse.org/images/logo.svg) Either use existing synapse client function or https://docs.synapse.org/rest/POST/externalFileHandle.html to create external file handle
2. make copies
You don't need to clean up these file handles -- these will be clean up every week when we relaunch the dev stack.

File: synapseutils/copy.py [hunk 6]
```diff
@@ -4,48 +4,75 @@
...
 import re
 import json
+import itertools
 ############################################################
 #                  Copy Functions                          #
 ############################################################
 
 
-def copyFileHandles(syn, fileHandles, associateObjectTypes, associateObjectIds, contentTypes, fileNames):
+def copyFileHandles(self, fileHandles, associateObjectTypes, associateObjectIds, newContentTypes=None, newFileNames=None):
```
[the remark this answers is not in the document]
Comment (kimyen) [reply synapseutils/copy.py]: Also, make the optional keyword arguments true keyword arguments by using this syntax: 
`def functionname(position_arg1, position_arg2, *, keyword_arg1=<default_value>, keyword_arg1=<default_value>):`

**Pull request closed without merging**